### ISSUE TITLE
"RelatedModel" has no attribute "RelatedObjectDoesNotExist"  [attr-defined]

### ISSUE THREAD

**Issue by eternal-sorrow** [typeddjango/django-stubs#1390]
# Bug report

## What's wrong

~Mypy prints `Module "django.db.models" has no attribute "ObjectDoesNotExist"  [attr-defined]` on the `from django.db.models import ObjectDoesNotExist` line, which is obviously wrong.~

Also, if I alternatively use `MyModel.field.RelatedObjectDoesNotExist`, I get the `"RelatedModel" has no attribute "RelatedObjectDoesNotExist"  [attr-defined]` message. 

`field` here is a reverse relation from `RelatedModel` to `MyModel` one-to-one field.

## How is that should be

No error messages.

## System information

- OS:
- `python` version: 3.10.6
- `django` version: 3.2.18
- `mypy` version: 1.0.1
- `django-stubs` version: 1.15.0
- `django-stubs-ext` version: 0.7.0

**Comment by intgr** [comment]
Hi. This was caused by https://github.com/typeddjango/django-stubs/pull/1309, which removes re-exports that are not part of Django's intentional public API.

According to Django documentation https://docs.djangoproject.com/en/4.1/ref/exceptions/ the exception `ObjectDoesNotExist` should be imported from `django.core.exceptions`, not from `django.db.models`.

**Comment by eternal-sorrow** [comment]
Alright, that's valid. The second prat of the bug report still applies.

**Comment by intgr** [comment]
Ah sorry I missed the second part. You're right, that's a bug. Looks like it could be complicated.

**Comment by RyanWalker277** [comment]
Hey! @eternal-sorrow can you share the code snippet in which you encountered the bug? On a quick lookup in `related_descriptors.pyi`

https://github.com/typeddjango/django-stubs/blob/0aa7bf1dc6effdc68b50e7c617ae2094e90a820c/django-stubs/db/models/fields/related_descriptors.pyi#L19-L34

https://github.com/typeddjango/django-stubs/blob/0aa7bf1dc6effdc68b50e7c617ae2094e90a820c/django-stubs/db/models/fields/related_descriptors.pyi#L40-L52

One thing that I think that might be causing the issue can be the fact that the `RelatedObjectDoesNotExist` attribute is defined as a property in the `ForwardManyToOneDescriptor` and `ReverseOneToOneDescriptor` classes, but it is not defined in the `ManyToManyDescriptor` class. This could result in the error when using `MyModel.field.RelatedObjectDoesNotExist` if field is a `ManyToManyField`. I may be wrong and something else might be causing the issue, just wanted to have a discussion on the potential causes before working on this one.

**Comment by eternal-sorrow** [comment]
`field` here is a reverse relation for a one-to-one field. I looks something like this:

```python
class A(models.Model):
    ...

class B(models.Model):
    a = models.OneToOneField(A, related_name='b')

a = A.objects.first()
try:
    b = a.b
except A.b.RelatedObjectDoesNotExist:  # mypy complains here '"B" has no attribute "RelatedObjectDoesNotExist"'
    pass
```

**Comment by flaeppe** [comment]
Closed via #1733
